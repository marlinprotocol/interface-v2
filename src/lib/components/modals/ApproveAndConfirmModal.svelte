<script lang="ts">
	import Button from '$lib/atoms/buttons/Button.svelte';
	import Modal from '$lib/atoms/modals/Modal.svelte';
	import Text from '$lib/atoms/texts/Text.svelte';
	import { DEFAULT_CURRENCY_DECIMALS } from '$lib/utils/constants/constants';
	import { bigNumberToString, mPondToPond, pondToMPond } from '$lib/utils/helpers/conversionHelper';
	import { closeModal, openModal } from '$lib/utils/helpers/commonHelper';
	import { staticImages } from '$lib/components/images/staticImages';
	import LoadingAnimationModal from '$lib/components/loading/LoadingAnimationModal.svelte';
	import { getAmountPrecision } from '$lib/utils/helpers/bridgeHelpers';

	export let handleApproveClick: () => Promise<void>;
	export let handleConfirmClick: () => Promise<void>;
	export let handleSuccessFinishClick: () => void;
	export let approved: boolean;
	export let rowIndex: number;
	export let approveButtonText = 'APPROVE';
	export let confirmButtonText = 'CONFIRM';
	export let amountConverted = 0n;
	export let conversionFrom: 'pond' | 'mPond' = 'pond';
	export let modalForApproveConfirm: string;

	let approveLoading: boolean;
	let confirmLoading: boolean;

	const approveClick = async () => {
		try {
			approveLoading = true;
			await handleApproveClick();
			approveLoading = false;
		} catch (e) {
			approveLoading = false;
			throw e;
		}
	};

	const confirmClick = async () => {
		try {
			confirmLoading = true;
			await handleConfirmClick();
			handleSuccessFinishClick();
			closeModal(modalForApproveConfirm);
			openModal(modalForSuccessConversion);
		} catch (e) {
			console.error(e);
			throw e;
		} finally {
			confirmLoading = false;
		}
	};
	// const modalWidth = 'max-w-[500px]';

	$: amountConvertedFrom = bigNumberToString(
		amountConverted,
		DEFAULT_CURRENCY_DECIMALS,
		getAmountPrecision(conversionFrom)
	);
	$: amountConvertedTo = bigNumberToString(
		conversionFrom === 'pond' ? pondToMPond(amountConverted) : mPondToPond(amountConverted),
		DEFAULT_CURRENCY_DECIMALS,
		getAmountPrecision(conversionFrom === 'pond' ? 'mPond' : 'pond')
	);
	$: conversionFromText = conversionFrom === 'pond' ? 'POND' : 'MPond';
	$: conversionToText = conversionFrom === 'pond' ? 'MPond' : 'POND';
	$: modalForSuccessConversion = `success-conversion-modal-${rowIndex}`;
</script>

<Modal modalFor={modalForApproveConfirm}>
	<svelte:fragment slot="title">
		{!approved ? 'Approve Transaction' : 'Confirm Transaction'}
	</svelte:fragment>
	<svelte:fragment slot="content">
		<div class="flex h-[50px] gap-5">
			<div class="flex flex-col items-center">
				{#if approved}
					<img src={staticImages.Check} alt="Copy" width="20px" height="20px" />
				{:else}
					<LoadingAnimationModal loading={true}>
						<Text variant="small" styleClass="font-semibold" text="1" />
					</LoadingAnimationModal>
				{/if}
				{#if approved}
					<div class="h-full w-[0.1px] bg-grey-400" />
				{/if}
			</div>
			<div class="grow text-left font-normal text-grey-500">
				Approve
				<span class="font-semibold text-black">
					{amountConvertedFrom}
					{conversionFromText}
				</span>
				for conversion
			</div>
		</div>
		<div class="flex gap-5">
			<LoadingAnimationModal loading={confirmLoading}>
				<Text variant="small" styleClass="font-semibold" text="2" />
			</LoadingAnimationModal>
			<div class="grow text-left font-normal {approved ? 'text-grey-500' : 'text-grey-300'}">
				Convert
				<span class={approved ? 'font-semibold text-black' : 'font-semibold text-grey-300'}>
					{amountConvertedFrom}
					{conversionFromText}
				</span>
				to
				<span class={approved ? 'font-semibold text-black' : 'font-semibold text-grey-300'}>
					{amountConvertedTo}
					{conversionToText}
				</span>
			</div>
		</div>
	</svelte:fragment>
	<svelte:fragment slot="actionButtons">
		{#if !approved}
			<Button
				onclick={approveClick}
				loading={approveLoading}
				variant="filled"
				size="large"
				styleClass="w-full"
			>
				{approveButtonText}
			</Button>
		{:else}
			<Button
				onclick={confirmClick}
				loading={confirmLoading}
				variant="filled"
				size="large"
				styleClass="w-full"
			>
				{confirmButtonText}
			</Button>
		{/if}
	</svelte:fragment>
</Modal>
