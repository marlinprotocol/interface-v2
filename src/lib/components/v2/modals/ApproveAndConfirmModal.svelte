<script lang="ts">
	import Button from '$lib/atoms/v2/buttons/Button.svelte';
	import Modal from '$lib/atoms/v2/modals/Modal.svelte';
	import SuccessfulConversionModal from '$lib/page-components/v2/bridge/modals/SuccessfulConversionModal.svelte';
	import { DEFAULT_CURRENCY_DECIMALS } from '$lib/utils/constants/constants';
	import { bigNumberToString, mPondToPond, pondToMPond } from '$lib/utils/helpers/conversionHelper';
	import { closeModal, cn, openModal } from '$lib/utils/helpers/commonHelper';
	import { staticImages } from '$lib/components/images/staticImages';
	import { removeTrailingZeros } from '$lib/utils/v2/helpers/commonHelper';

	export let handleApproveClick: () => Promise<void>;
	export let handleConfirmClick: () => Promise<void>;
	export let handleSuccessFinishClick: () => void;
	export let approved: boolean;
	export let rowIndex: number;
	export let approveButtonText = 'APPROVE';
	export let confirmButtonText = 'CONFIRM';
	export let amountConverted = 0n;
	export let conversionFrom: 'pond' | 'mPond' = 'pond';
	export let modalForApproveConfirm: string;

	let approveLoading: boolean;
	let confirmLoading: boolean;

	const approveClick = async () => {
		try {
			approveLoading = true;
			await handleApproveClick();
			approveLoading = false;
		} catch (e) {
			approveLoading = false;
			throw e;
		}
	};

	const confirmClick = async () => {
		try {
			confirmLoading = true;
			await handleConfirmClick();
			handleSuccessFinishClick();
			closeModal(modalForApproveConfirm);
			openModal(modalForSuccessConversion);
		} catch (e) {
			console.error(e);
			throw e;
		} finally {
			confirmLoading = false;
		}
	};
	// const modalWidth = 'max-w-[500px]';

	$: amountConvertedFrom = removeTrailingZeros(
		bigNumberToString(amountConverted, DEFAULT_CURRENCY_DECIMALS, 18)
	);
	$: amountConvertedTo = removeTrailingZeros(
		bigNumberToString(
			conversionFrom === 'pond' ? pondToMPond(amountConverted) : mPondToPond(amountConverted),
			DEFAULT_CURRENCY_DECIMALS,
			18
		)
	);
	$: conversionFromText = conversionFrom === 'pond' ? 'POND' : 'MPond';
	$: conversionToText = conversionFrom === 'pond' ? 'MPond' : 'POND';
	$: modalForSuccessConversion = `success-conversion-modal-${rowIndex}`;
</script>

<Modal modalFor={modalForApproveConfirm}>
	<svelte:fragment slot="title">
		{!approved ? 'Approve Transaction' : 'Confirm Transaction'}
	</svelte:fragment>
	<svelte:fragment slot="content">
		<div class="mb-2 flex h-20 items-center gap-3 rounded-xl bg-[#F7F7F7] p-3 font-poppins">
			<div class="flex flex-col items-center">
				<div
					class={cn(
						'flex h-14 w-14 items-center justify-center rounded-full text-xl font-semibold text-white',
						{
							'bg-[#68A843]': approved,
							'bg-[#EDBE59]': !approved
						}
					)}
				>
					{#if approved}
						<img src={staticImages.checkmark} alt="checkmark" width="25px" height="25px" />
					{:else}
						1
					{/if}
				</div>
			</div>
			<div class="grow text-left text-xl font-normal text-[#030115]">
				Approve
				<span class="font-semibold">
					{amountConvertedFrom}
					{conversionFromText}
				</span>
				for conversion
			</div>
		</div>
		<div
			class={cn('flex h-20 items-center gap-3 rounded-xl p-3 font-poppins', {
				'bg-[#F7F7F7]': approved,
				'border border-[#D9DADE]': !approved
			})}
		>
			<div
				class={cn(
					'flex h-14 w-14 items-center justify-center rounded-full text-xl font-semibold text-white',
					{
						'bg-[#EDBE59]': approved,
						'bg-[#F7F7F7] text-[#939393]': !approved
					}
				)}
			>
				2
			</div>
			<div class="grow text-left text-xl font-normal text-[#030115]">
				Convert
				<span class="font-semibold">
					{amountConvertedFrom}
					{conversionFromText}
				</span>
				to
				<span class="font-semibold">
					{amountConvertedTo}
					{conversionToText}
				</span>
			</div>
		</div>
	</svelte:fragment>
	<svelte:fragment slot="actionButtons">
		{#if !approved}
			<Button
				onclick={approveClick}
				loading={approveLoading}
				variant="filled"
				size="large"
				styleClass="w-full"
			>
				{approveButtonText}
			</Button>
		{:else}
			<Button
				onclick={confirmClick}
				loading={confirmLoading}
				variant="filled"
				size="large"
				styleClass="w-full"
			>
				{confirmButtonText}
			</Button>
		{/if}
	</svelte:fragment>
</Modal>
<SuccessfulConversionModal
	modalFor={modalForSuccessConversion}
	{amountConverted}
	{conversionFrom}
/>
