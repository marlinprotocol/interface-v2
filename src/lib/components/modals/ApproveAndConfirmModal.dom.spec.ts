import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent } from '@testing-library/svelte';
import ApproveAndConfirmModal from './ApproveAndConfirmModal.svelte';
import { tick } from 'svelte';

describe('ApproveAndConfirmModal', () => {
	it("render's the ApproveAndConfirmModal properly", () => {
		const handleApproveClick = vi.fn().mockImplementation(() => Promise.resolve());
		const handleConfirmClick = vi.fn().mockImplementation(() => Promise.resolve());
		const { container } = render(ApproveAndConfirmModal, {
			handleApproveClick,
			approved: false,
			handleConfirmClick,
			rowIndex: 10,
			modalForApproveConfirm: 'test-id-some-id',
			handleSuccessFinishClick: () => {},
			conversionFrom: 'mPond',
			amountConverted: 123999999999n,
			confirmButtonText: 'CONVERT'
		});
		expect(container).toMatchSnapshot();
	});
	it('should call handleApproveClick when approve button is clicked and render relevant elements', async () => {
		const handleApproveClick = vi.fn().mockImplementation(() => Promise.resolve());
		const handleConfirmClick = vi.fn().mockImplementation(() => Promise.resolve());
		const { getByText } = render(ApproveAndConfirmModal, {
			handleApproveClick,
			approved: false,
			handleConfirmClick,
			rowIndex: 10,
			modalForApproveConfirm: 'test-id-some-id',
			handleSuccessFinishClick: () => {},
			conversionFrom: 'mPond',
			amountConverted: 123999999999n,
			confirmButtonText: 'CONVERT'
		});
		expect(getByText('Approve Transaction')).toBeTruthy();
		await fireEvent.click(getByText('APPROVE'));
		await tick();
		expect(handleApproveClick).toHaveBeenCalledOnce();
	});
	it('should call handleConfirmClick when confirm button is clicked and render relevant elements', async () => {
		const handleConfirmClick = vi.fn().mockImplementation(() => Promise.resolve());
		const handleSuccessFinishClick = vi.fn();
		const handleApproveClick = vi.fn().mockImplementation(() => Promise.resolve());

		const { getByText } = render(ApproveAndConfirmModal, {
			handleApproveClick,
			approved: true,
			handleConfirmClick,
			rowIndex: 10,
			modalForApproveConfirm: 'test-id-some-id',
			handleSuccessFinishClick,
			conversionFrom: 'mPond',
			amountConverted: 123999999999n,
			confirmButtonText: 'CONVERT'
		});

		expect(getByText('Confirm Transaction')).toBeTruthy();
		await fireEvent.click(getByText('CONVERT'));

		expect(handleConfirmClick).toHaveBeenCalled();
		await tick();
		expect(handleSuccessFinishClick).toHaveBeenCalled();
	});
});
