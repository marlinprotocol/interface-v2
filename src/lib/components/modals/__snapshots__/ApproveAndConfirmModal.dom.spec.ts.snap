// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ApproveAndConfirmModal > render's the ApproveAndConfirmModal properly 1`] = `
<body>
  <onboard-v2
    style="all: initial;"
  />
  <style>
    
    
@font-face {
  font-family: 'Inter';
  font-style:  normal;
  font-weight: 300 600;
  font-display: swap;
  src: url("https://rsms.me/inter/font-files/Inter-roman.var.woff2?v=3.19") format("woff2");
}

  
  </style>
  <div>
    <input
      class="modal-toggle s-GZwqZrPaBy6h"
      id="test-id-some-id"
      tabindex="-1"
      type="checkbox"
    />
     
    <div
      class="modal s-GZwqZrPaBy6h"
      data-testid="modal"
    >
      <div
        class="modal-box flex flex-col rounded-3xl bg-[#FCFCFC] p-0 drop-shadow-sm w-11/12 sm:w-3/4 sm:max-w-[607px] overflow-x-hidden overflow-y-visible s-GZwqZrPaBy6h"
      >
        <div
          class="modal-header sticky top-0 z-10 flex items-center justify-between bg-[#FCFCFC] px-6 pt-6 s-GZwqZrPaBy6h"
        >
          <div
            class="flex w-full flex-col s-GZwqZrPaBy6h"
          >
             
             
             
            <div
              class="text-left text-[22px] font-medium tracking-[-1px] s-GZwqZrPaBy6h"
            >
              Approve Transaction
            </div>
             
          </div>
           
          <button
            class="flex h-[54px] w-[54px] shrink-0 cursor-pointer items-center justify-center self-baseline rounded-full border border-[#D9DADE] bg-white text-sm font-bold text-gray-300 s-GZwqZrPaBy6h"
            data-testid="modal-close-button"
          >
            <img
              alt="Close"
              class="s-GZwqZrPaBy6h"
              src="/images/closeIcon.svg"
            />
          </button>
        </div>
         
        <div
          class="modal-body px-6 pt-4 s-GZwqZrPaBy6h"
        >
          <div
            class="mb-2 flex h-20 items-center gap-3 rounded-xl bg-[#F7F7F7] p-3 font-poppins"
          >
            <div
              class="flex flex-col items-center"
            >
              <div
                class="flex h-14 w-14 items-center justify-center rounded-full text-xl font-semibold text-white bg-[#EDBE59]"
              >
                1
              </div>
            </div>
             
            <div
              class="grow text-left text-xl font-normal text-[#030115]"
            >
              Approve
				
              <span
                class="font-semibold"
              >
                0.000000123999999999
                 
                MPond
              </span>
              
				for conversion
            </div>
          </div>
           
          <div
            class="flex h-20 items-center gap-3 rounded-xl p-3 font-poppins border border-[#D9DADE]"
          >
            <div
              class="flex h-14 w-14 items-center justify-center rounded-full text-xl font-semibold bg-[#F7F7F7] text-[#939393]"
            >
              2
            </div>
             
            <div
              class="grow text-left text-xl font-normal text-[#030115]"
            >
              Convert
				
              <span
                class="font-semibold"
              >
                0.000000123999999999
                 
                MPond
              </span>
              
				to
				
              <span
                class="font-semibold"
              >
                0.123999999999
                 
                POND
              </span>
            </div>
          </div>
        </div>
         
        <div
          class="modal-footer sticky bottom-0 z-[2] bg-[#FCFCFC] px-6 pb-6 pt-4 s-GZwqZrPaBy6h"
        >
          <button
            class="gap-1.5 h-14 text-base font-normal btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed !rounded-[100px] btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white w-full"
            id=""
          >
             
            APPROVE
          </button>
          
          <!--&lt;Button&gt;-->
          
        </div>
         
      </div>
    </div>
    <!--&lt;Modal&gt;-->
     
    <input
      class="modal-toggle s-GZwqZrPaBy6h"
      id="success-conversion-modal-10"
      tabindex="-1"
      type="checkbox"
    />
     
    <div
      class="modal s-GZwqZrPaBy6h"
      data-testid="modal"
    >
      <div
        class="modal-box flex flex-col rounded-3xl bg-[#FCFCFC] p-0 drop-shadow-sm w-11/12 sm:w-3/4 sm:max-w-[607px] overflow-x-hidden overflow-y-visible s-GZwqZrPaBy6h"
      >
        <div
          class="modal-header sticky top-0 z-10 flex justify-between bg-[#FCFCFC] px-6 pt-6 items-start s-GZwqZrPaBy6h"
        >
          <div
            class="flex w-full flex-col s-GZwqZrPaBy6h"
          >
             
            <div
              class="flex w-full flex-col items-center justify-between s-GZwqZrPaBy6h"
            >
              <div
                class="ml-[70px] grid h-[84px] w-[84px] place-items-center rounded-full bg-[#68A843] s-GZwqZrPaBy6h"
              >
                <img
                  alt="tick"
                  class="s-GZwqZrPaBy6h"
                  height="60px"
                  src="/images/tick.svg"
                  width="60px"
                />
              </div>
               
              <div
                class="ml-[70px] mt-4 text-[26px] font-semibold s-GZwqZrPaBy6h"
              >
                Conversion Successful
              </div>
            </div>
             
             
             
          </div>
           
          <button
            class="flex h-[54px] w-[54px] shrink-0 cursor-pointer items-center justify-center self-baseline rounded-full border border-[#D9DADE] bg-white text-sm font-bold text-gray-300 s-GZwqZrPaBy6h"
            data-testid="modal-close-button"
          >
            <img
              alt="Close"
              class="s-GZwqZrPaBy6h"
              src="/images/closeIcon.svg"
            />
          </button>
        </div>
         
        <div
          class="modal-body px-6 pt-4 s-GZwqZrPaBy6h"
        >
          <div
            class="flex flex-col gap-6 text-lg font-light text-[#26272C]"
          >
            <div
              class="whitespace-normal rounded-xl border border-[#D9DADE] p-4"
            >
              <div
                class="mb-2 text-center"
              >
                You have converted
              </div>
               
              <div
                class="text-center"
              >
                <span
                  class="font-bold text-black"
                >
                  0.000000123999999999
                   
                  MPOND
                </span>
                
					to
					
                <span
                  class="font-bold text-black"
                >
                  0.123999999999
                   
                  POND
                </span>
              </div>
            </div>
             
            <div
              class="whitespace-normal rounded-xl border border-[#D9DADE] p-4"
            >
              <div
                class="mb-2 text-center"
              >
                Updated Wallet Balance
              </div>
               
              <p
                class="text-center font-bold text-black"
              >
                0
                
					POND
              </p>
               
              <p
                class="text-center font-bold text-black"
              >
                0
                 MPond
              </p>
            </div>
          </div>
        </div>
         
        <div
          class="modal-footer sticky bottom-0 z-[2] bg-[#FCFCFC] px-6 pb-6 pt-4 s-GZwqZrPaBy6h"
        >
          <button
            class="gap-1.5 h-14 text-base font-normal btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed !rounded-[100px] btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white w-full"
            id=""
          >
             
            Finish
          </button>
          
          <!--&lt;Button&gt;-->
        </div>
         
      </div>
    </div>
    <!--&lt;Modal&gt;-->
    <!--&lt;SuccessfulConversionModal&gt;-->
    <!--&lt;ApproveAndConfirmModal&gt;-->
  </div>
</body>
`;
